### PR TITLE
TRITON-2010 Triton components should be aware of CNs encryption (EDAR)

### PR DESCRIPTION
**Pull request by kusor**
Not before https://github.com/joyent/sdc-vmapi/pull/53

### CODE REVIEW

**Comment by twhiteman** [comment]
I can review once discussion for https://github.com/joyent/sdc-vmapi/pull/53#discussion_r358381125 is done.

**Comment by kusor** [comment]
> I can review once discussion for [joyent/sdc-vmapi#53 (comment)](https://github.com/joyent/sdc-vmapi/pull/53#discussion_r358381125) is done.

Updated to use the new `internal_metadata.encrypted` instead of just `encrypted` for messages going to VMAPI.

File: test/getCreateOptions.test.js [hunk 1]
```diff
@@ -205,14 +211,34 @@ test('getCreateOptions sets brand to "joyent" by default', function (t) {
 
     createOpts = getCreateOptions(req);
 
-    t.equal(createOpts.brand, 'joyent', 'default brand should be joyent');
+    t.equal(createOpts.brand, req.config.default_container_brand,
+        'default brand should be joyent');
```
Comment (twhiteman) [inline test/getCreateOptions.test.js]: Maybe won't be "joyent" - so just say "check container brand is the the default" ?

File: test/getCreateOptions.test.js [hunk 2]
```diff
@@ -205,14 +211,34 @@ test('getCreateOptions sets brand to "joyent" by default', function (t) {
...
+        img: IMAGES['smartos-1.6.3'],
+        params: {
+            encrypted: true
+        }
+    });
+
+    createOpts = getCreateOptions(req);
+
+    t.equal(createOpts.brand, req.config.default_container_brand,
+        'default brand should be joyent');
```
Comment (twhiteman) [inline test/getCreateOptions.test.js]: "joyent" again

File: test/getCreateOptions.test.js [hunk 3]
```diff
@@ -222,14 +248,14 @@ test('getCreateOptions sets brand to "kvm" by when img.type === zvol',
 
     createOpts = getCreateOptions(req);
 
-    t.equal(createOpts.brand, 'kvm', 'default brand should be kvm for zvol');
+    t.equal(createOpts.brand, req.config.default_hvm_brand,
+        'default brand should be kvm for zvol');
```
Comment (twhiteman) [inline test/getCreateOptions.test.js]: "kvm"

File: lib/machines.js [hunk 4]
```diff
@@ -279,6 +278,10 @@ function translate(machine, req) {
         });
     }
 
+    if (machine.internal_metadata && machine.internal_metadata.encrypted) {
+        msg.encrypted = machine.internal_metadata.encrypted;
```
Comment (twhiteman) [inline lib/machines.js]: Just checking - we leave out the "encrypted" property when the vm is not encrypted (i.e. this property is optional).

File: lib/machines.js [hunk 5]
```diff
@@ -610,6 +613,11 @@ function getCreateOptions(req) {
         opts.indestructible_zoneroot = true;
     }
 
+
+    if (params.encrypted === true) {
+        opts.encrypted = true;
```
Comment (twhiteman) [inline lib/machines.js]: This should be setting 'opts.internal_metadata.encrypted' right?

File: docs/index.md [hunk 6]
```diff
@@ -4809,6 +4813,7 @@ deletion_protection | Boolean | Whether an instance is destroyable. See [Deletio
 allow_shared_images | Boolean | Whether to allow provisioning from a shared image. Default is false
 volumes   | Array    | A list of objects representing volumes to mount when the newly created machine boots
 disks     | Array    | An array of disk objects to be created (bhyve)
+encrypted | Boolean  | Place this instance into an encrypted server. Default is false
```
Comment (twhiteman) [inline docs/index.md]: Add "optional" wording (if you note which properties are optional).

File: lib/machines.js (hunk 5 above)
Comment (twhiteman) [reply lib/machines.js]: Oh, it's in the next commit.

File: lib/machines.js [hunk 7]
```diff
@@ -858,8 +854,20 @@ function getCreateOptions(req) {
...
+        if (!opts.internal_metadata) {
+            opts.internal_metadata = {};
+        }
+        opts.internal_metadata.encrypted = true;
+    }
+
+    if (opts.internal_metadata) {
         // Pass a string to VMAPI, not an object, or validation will fail
-        opts.internal_metadata = JSON.stringify(credentials);
+        opts.internal_metadata = JSON.stringify(opts.internal_metadata);
```
Comment (twhiteman) [inline lib/machines.js]: Hmmm... this doesn't seem right. The `internal_metadata` property should be passed as an object.

Comment (kusor) [reply lib/machines.js]: It's lame, but it's how it works: https://github.com/joyent/sdc-vmapi/blob/master/lib/common/validation.js#L1755-L1758

Some testing, with this CloudAPI's code. These are Job's parameters:
```
  "internal_metadata": {
    "root_pw": "secret",
    "admin_pw": "secret",
    "encrypted": true,
    "locality": "{\"far\":\"af4167f0-beda-4af9-9ae4-99d544499c14\",\"strict\":true}"
  },
```

And it of course went into an encrypted server.

File: docs/index.md (hunk 6 above)
Comment (kusor) [reply docs/index.md]: Done

File: lib/machines.js (hunk 4 above)
Comment (kusor) [reply lib/machines.js]: Yeah, it just matter when it's encrypted. Whatever else doesn't really matter

File: test/getCreateOptions.test.js (hunk 1 above)
Comment (kusor) [reply test/getCreateOptions.test.js]: Done!

File: test/getCreateOptions.test.js (hunk 2 above)
Comment (kusor) [reply test/getCreateOptions.test.js]: Done!

File: test/getCreateOptions.test.js (hunk 3 above)
Comment (kusor) [reply test/getCreateOptions.test.js]: Done!

File: lib/machines.js (hunk 7 above)
Comment (twhiteman) [reply lib/machines.js]: That code (linked above) is only for the vms.list endpoint - not for vms.create endpoint.

It looks like this is the code that does the string to object transformation for internal_metadata during vm.create:
https://github.com/joyent/sdc-vmapi/blob/35fbeaabd424ccf7b2d4c6d305789c5762b7b8b3/lib/common/validation.js#L2455

I don't know why one technique complains and the other does not - that seems like a validation bug (in vmapi) to me.

**Pull request merged**